File: backend/services/post-ts-koa-mongo-fdm/src/edit-post/impl.ts
import { Deps, now } from "../common/utils";
import { Workflow } from ".";
import { PostId, UserId, PostContent } from "../common";
import * as db from "../common/utils/db";
import { Long, ObjectId, WithId } from "mongodb";
import { throwUnexpectedValue } from "../common/utils/error";
import { errors } from "./api";
import * as runtypes from "runtypes";

export class WorkflowImpl extends Workflow {
    private static readonly postHasCreator_TextOrUrl = runtypes.Record({
        creator: runtypes.InstanceOf(ObjectId),
        text: runtypes.String.optional(),
        url: runtypes.String.optional(),
    });

    async checkUserIsCreatorAndContentHasTheSameType(postId: PostId, userId: UserId, content: PostContent): Promise<void> {
        const postOid = db.tryParseId(postId);
        if (postOid === undefined) {
            throw this.errors.postNotFound();
        }
        const post: WithId<unknown> | null = await this.deps.mongoDb.collection(db.posts).findOne({ _id: postOid });
        if (post === null) {
            throw this.errors.postNotFound();
        }
        db.validate(WorkflowImpl.postHasCreator_TextOrUrl, db.posts, post);
        
        const userOid = db.tryParseId(userId);
        if (userOid === undefined) {
            throw this.errors.notCreator();
        }
        if (!post.creator.equals(userOid)) {
            throw this.errors.notCreator();
        }
        const sameType =
            content.type === "Text" ? post.text !== undefined :
                content.type === "Url" ? post.url !== undefined :
                    throwUnexpectedValue(content);
        if (!sameType) {
            throw this.errors.typeDiff();
        }
    }

    async updatePost(postId: PostId, newContent: PostContent): Promise<void> {
        const postOid = db.tryParseId(postId);
        if (postOid === undefined) {
            throw this.errors.postNotFound();
        }

        const lastModified = Long.fromNumber(now().utc);
        const updateSet: {
            lastModified: Long,
        } & ({
            text: string
        } | {
            url: string
        }) = newContent.type === "Text" ? { lastModified, text: newContent.content } :
                newContent.type === "Url" ? { lastModified, url: newContent.content } :
                    throwUnexpectedValue(newContent);
        await this.deps.mongoDb.collection(db.posts).updateOne({ _id: postOid }, { $set: updateSet });
    }

    readonly errors = errors;
    constructor(private deps: Deps) { super(); }
}

File: backend/services/post-ktor-mongo-fdm/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/edit_post/impl.kt
package me.jason5lee.post_ktor_mongo_fdm.edit_post

import com.mongodb.client.model.Projections
import kotlinx.coroutines.reactive.awaitSingle
import me.jason5lee.post_ktor_mongo_fdm.common.PostContent
import me.jason5lee.post_ktor_mongo_fdm.common.PostId
import me.jason5lee.post_ktor_mongo_fdm.common.UserId
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps
import me.jason5lee.post_ktor_mongo_fdm.common.utils.timeFromNow
import me.jason5lee.resukt.Result
import org.bson.Document
import org.bson.types.ObjectId

class WorkflowImpl(private val deps: Deps) : Workflow() {
    override suspend fun checkUserIsCreatorAndContentHasTheSameType(
        post: PostId,
        user: UserId,
        content: PostContent
    ): Result<Unit, Errors> {
        val postDoc = Db.findById(
            db = deps.mongoDb,
            collection = Db.posts,
            id = Db.toObjectIdOrNull(post.value) ?: return Result.failure(Errors.PostNotFound),
            projection = Projections.include("creator", "text", "url")
        ) ?: return Result.failure(Errors.PostNotFound)
        if (postDoc.get<ObjectId>("creator") != Db.toObjectIdOrNull(user.value)) {
            return Result.failure(Errors.NotCreator)
        }
        val sameType = when (content) {
            is PostContent.Text -> postDoc.doc.containsKey("text")
            is PostContent.Url -> postDoc.doc.containsKey("url")
        }
        if (!sameType) {
            return Result.failure(Errors.TypeDiff)
        }
        return Result.success(Unit)
    }

    override suspend fun updatePost(post: PostId, newContent: PostContent): Result<Unit, Errors> {
        val lastModified = timeFromNow(offset = 0).utc
        val updateSet = when (newContent) {
            is PostContent.Text -> Document("text", newContent.value.value)
            is PostContent.Url -> Document("url", newContent.value.value)
        }
        updateSet.append("lastModified", lastModified)
        deps.mongoDb.getCollection(Db.posts).updateOne(
            Document("_id", Db.toObjectIdOrNull(post.value) ?: return Result.failure(Errors.PostNotFound)),
            Document("\$set", updateSet)
        ).awaitSingle()
        return Result.success(Unit)
    }
}
